### PR TITLE
Fix staticcheck for integration test framework, examples, and garbage…

### PR DESCRIPTION
**Pull request by helenfeng737**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
 /kind cleanup

**What this PR does / why we need it**:
Part of https://github.com/kubernetes/kubernetes/issues/92402

**Which issue(s) this PR fixes**:
Part of issue: https://github.com/kubernetes/kubernetes/issues/92402

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/integration/examples/apiserver_test.go [hunk 1]
```diff
@@ -90,7 +90,7 @@ func TestAggregatedAPIServer(t *testing.T) {
 			"--kubeconfig", wardleToKASKubeConfigFile,
 		})
 		if err := wardleCmd.Execute(); err != nil {
-			t.Fatal(err)
+			t.Errorf("unexpected error: %v", err)
```
Comment (neolit123) [inline test/integration/examples/apiserver_test.go:93]: could you explain this change?

File: test/integration/framework/test_server.go [hunk 2]
```diff
@@ -128,7 +128,7 @@ func StartTestServer(t *testing.T, stopCh <-chan struct{}, setup TestServerSetup
 	}
 	go func() {
 		if err := kubeAPIServer.GenericAPIServer.PrepareRun().Run(stopCh); err != nil {
-			t.Fatal(err)
+			t.Errorf("unexpected error: %v", err)
```
Comment (neolit123) [inline test/integration/framework/test_server.go:131]: this too.

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 3]
```diff
@@ -443,7 +443,7 @@ func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet
 	rc.ObjectMeta.Finalizers = initialFinalizers
 	rc, err := rcClient.Create(context.TODO(), rc, metav1.CreateOptions{})
 	if err != nil {
-		t.Fatalf("Failed to create replication controller: %v", err)
+		t.Errorf("Failed to create replication controller: %v", err)
```
Comment (neolit123) [inline test/integration/garbagecollector/garbage_collector_test.go:446]: [1]

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 4]
```diff
@@ -453,18 +453,15 @@ func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet
 		pod := newPod(podName, namespace, []metav1.OwnerReference{{UID: rc.ObjectMeta.UID, Name: rc.ObjectMeta.Name}})
 		createdPod, err := podClient.Create(context.TODO(), pod, metav1.CreateOptions{})
 		if err != nil {
-			t.Fatalf("Failed to create Pod: %v", err)
+			t.Errorf("Failed to create Pod: %v", err)
```
Comment (neolit123) [inline test/integration/garbagecollector/garbage_collector_test.go:456]: [1]

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 5]
```diff
@@ -485,12 +482,12 @@ func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet
 			return true, nil
 		})
 		if err != nil {
-			t.Fatalf("failed to observe the expected pods in the GC graph for rc %s", rcName)
+			t.Errorf("failed to observe the expected pods in the GC graph for rc %s", rcName)
```
Comment (neolit123) [inline test/integration/garbagecollector/garbage_collector_test.go:485]: [1]

File: test/integration/examples/apiserver_test.go (hunk 1 above)
Comment (ZhiFeng1993) [reply test/integration/examples/apiserver_test.go:93]: I got this: 
```
Errors from staticcheck:
test/integration/examples/apiserver_test.go:80:2: the goroutine calls T.Fatalf, which must be called in the same goroutine as the test (SA2002)
test/integration/examples/apiserver_test.go:93:4: call to T.Fatalf
```
https://github.com/kubernetes/kubernetes/blob/master/test/integration/examples/apiserver_test.go#L78

This function is used by some other function so I assume changing this will not change the test behavior?

Comment (neolit123) [reply test/integration/examples/apiserver_test.go:93]: having another look,
the correct fix in such a case is to capture the potential error from the goroutine and call t.Fatal in the main thread.

@kubernetes/sig-api-machinery-pr-reviews

Comment (ZhiFeng1993) [reply test/integration/examples/apiserver_test.go:93]: @neolit123 thanks for the advice. It makes sense to me. I think it is a better fix. Will update

**Comment by neolit123** [comment]
the latest update seems fine, overall.
i think given this touches api-server related integration tests it should get review from @kubernetes/sig-api-machinery-pr-reviews 
if nobody is responding try asking in #sig-api-machinery on k8s slack.

also we are in code freeze for 1.19 and this should merge after code freeze is lifted.

**Comment by ZhiFeng1993** [comment]
> the latest update seems fine, overall.
> i think given this touches api-server related integration tests it should get review from @kubernetes/sig-api-machinery-pr-reviews
> if nobody is responding try asking in #sig-api-machinery on k8s slack.
> 
> also we are in code freeze for 1.19 and this should merge after code freeze is lifted.

Will do. Thanks for the review.

File: test/integration/examples/apiserver_test.go [hunk 6]
```diff
@@ -90,9 +92,14 @@ func TestAggregatedAPIServer(t *testing.T) {
 			"--kubeconfig", wardleToKASKubeConfigFile,
 		})
 		if err := wardleCmd.Execute(); err != nil {
-			t.Fatal(err)
+			errs <- err
```
Comment (knight42) [inline test/integration/examples/apiserver_test.go:95]: I think you need to `close(errs)`  if no error occurs, otherwise `err = <-errs` would block indefinitely.

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 7]
```diff
@@ -443,7 +443,7 @@ func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet
 	rc.ObjectMeta.Finalizers = initialFinalizers
 	rc, err := rcClient.Create(context.TODO(), rc, metav1.CreateOptions{})
 	if err != nil {
-		t.Fatalf("Failed to create replication controller: %v", err)
+		errs <- err
```
Comment (knight42) [inline test/integration/garbagecollector/garbage_collector_test.go:446]:
```suggestion
		errs <- fmt.Errorf("Failed to create replication controller: %v", err)
```
it might be better to perserve the original error message, like

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 8]
```diff
@@ -433,7 +433,7 @@ func TestCreateWithNonExistentOwner(t *testing.T) {
 	}
 }
 
-func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet clientset.Interface, nameSuffix, namespace string, initialFinalizers []string, options metav1.DeleteOptions, wg *sync.WaitGroup, rcUIDs chan types.UID) {
+func setupRCsPods(t *testing.T, gc *garbagecollector.GarbageCollector, clientSet clientset.Interface, nameSuffix, namespace string, initialFinalizers []string, options metav1.DeleteOptions, wg *sync.WaitGroup, rcUIDs chan types.UID, errs chan error) {
```
Comment (knight42) [inline test/integration/garbagecollector/garbage_collector_test.go:436]: again, `close(errs)` if no error occurs

File: test/integration/examples/apiserver_test.go [hunk 9]
```diff
@@ -131,7 +138,7 @@ func TestAggregatedAPIServer(t *testing.T) {
 
 	// wait for the unavailable API service to be processed with updated status
 	err = wait.Poll(100*time.Millisecond, 5*time.Second, func() (done bool, err error) {
-		_, err = kubeClient.Discovery().ServerResources()
+		_, _, err = kubeClient.Discovery().ServerGroupsAndResources()
```
Comment (liggitt) [inline test/integration/examples/apiserver_test.go:141]: why is this changed?

File: test/integration/garbagecollector/garbage_collector_test.go [hunk 10]
```diff
@@ -534,16 +531,43 @@ func TestStressingCascadingDeletion(t *testing.T) {
 	wg.Add(collections * 5)
 	rcUIDs := make(chan types.UID, collections*5)
 	for i := 0; i < collections; i++ {
+		// use a channel to hand off the error
 		// rc is created with empty finalizers, deleted with nil delete options, pods will remain.
-		go setupRCsPods(t, gc, clientSet, "collection1-"+strconv.Itoa(i), ns.Name, []string{}, metav1.DeleteOptions{}, &wg, rcUIDs)
+		errs := make(chan error, 1)
+		go setupRCsPods(t, gc, clientSet, "collection1-"+strconv.Itoa(i), ns.Name, []string{}, metav1.DeleteOptions{}, &wg, rcUIDs, errs)
+		err := <-errs
```
Comment (liggitt) [inline test/integration/garbagecollector/garbage_collector_test.go:538]: blocking on receiving the error means these all run serially, which defeats the purpose of the test. it would be simpler to change the Fatal to an Error

File: test/integration/framework/test_server.go [hunk 11]
```diff
@@ -126,11 +126,18 @@ func StartTestServer(t *testing.T, stopCh <-chan struct{}, setup TestServerSetup
...
+	// use a channel to hand off the error
+	errs := make(chan error, 1)
 	go func() {
 		if err := kubeAPIServer.GenericAPIServer.PrepareRun().Run(stopCh); err != nil {
-			t.Fatal(err)
+			errs <- err
 		}
 	}()
+	// wait for it
+	err = <-errs
```
Comment (liggitt) [inline test/integration/framework/test_server.go:137]: blocking here until the server stops defeats the purpose of starting it in a goroutine

File: test/integration/examples/apiserver_test.go [hunk 12]
```diff
@@ -90,9 +92,14 @@ func TestAggregatedAPIServer(t *testing.T) {
 			"--kubeconfig", wardleToKASKubeConfigFile,
 		})
 		if err := wardleCmd.Execute(); err != nil {
-			t.Fatal(err)
+			errs <- err
 		}
 	}()
+	// wait for it
+	err = <-errs
```
Comment (liggitt) [inline test/integration/examples/apiserver_test.go:99]: blocking here until the server stops defeats the purpose of starting it in a goroutine

File: test/integration/garbagecollector/garbage_collector_test.go (hunk 10 above)
Comment (ZhiFeng1993) [reply test/integration/garbagecollector/garbage_collector_test.go:538]: Will do.

File: test/integration/examples/apiserver_test.go (hunk 6 above)
Comment (ZhiFeng1993) [reply test/integration/examples/apiserver_test.go:95]: Thanks. I'll remove the channel and change it to err per @liggitt's suggestion.

File: test/integration/examples/apiserver_test.go (hunk 12 above)
Comment (ZhiFeng1993) [reply test/integration/examples/apiserver_test.go:99]: Will remove the channel and change it to regular error message.

File: test/integration/examples/apiserver_test.go (hunk 9 above)
Comment (ZhiFeng1993) [reply test/integration/examples/apiserver_test.go:141]: This is a static check error. `ServerResources` has been deprecated so we'd use `ServerGroupsAndResources` instead.

File: test/integration/framework/test_server.go (hunk 11 above)
Comment (ZhiFeng1993) [reply test/integration/framework/test_server.go:137]: Will change back to t.Error

File: test/integration/garbagecollector/garbage_collector_test.go (hunk 8 above)
Comment (ZhiFeng1993) [reply test/integration/garbagecollector/garbage_collector_test.go:436]: Will change back to t.Error
